### PR TITLE
feat(CC-0075): Add topologySpreadConstraints and PriorityClass support

### PR DESCRIPTION
**Pull request by berendt**
# GitHub Issue #207: Add topologySpreadConstraints and PriorityClass support to the Keystone Deployment

**Category**: improvement | **Scope**: Small

**Description**: `buildKeystoneDeployment` (operators/keystone/internal/controller/reconcile_deployment.go:150) currently produces a Deployment whose pod template has neither `topologySpreadConstraints`, nor `affinity`, nor `priorityClassName`. With the default `replicas: 3` plus the `minAvailable: 1` PDB (CC-0037, reconcile_deployment.go:284-307), nothing prevents the scheduler from coalescing all three Keystone pods onto the same node, so a single node reboot, kernel panic, or kubelet restart still produces a total Keystone outage despite the apparent 3-way HA. Add (1) sensible default `TopologySpreadConstraints` to the generated pod template — `maxSkew: 1` across `topology.kubernetes.io/zone` and `kubernetes.io/hostname`, both with `whenUnsatisfiable: ScheduleAnyway` so that the defaults never deadlock single-node dev clusters — and (2) an optional `spec.priorityClassName` field on the Keystone CR that is wired through to the pod template when set. Both should be overridable via new optional fields on the CRD: `spec.topologySpreadConstraints` and `spec.priorityClassName`.

**Motivation**: Keystone is the gating dependency for every other OpenStack service, so an unavailable Keystone takes down the entire control plane. The current setup looks HA on paper (3 replicas + PDB) but provides no anti-coalescing guarantees: a freshly-scheduled CR on a small cluster regularly lands all three pods on the same node, and even on multi-node clusters the scheduler has no signal to spread them. PriorityClass support is needed so operators can mark Keystone with a high-priority class on saturated control-plane nodes, ensuring it wins eviction races against tenant workloads. The change is intentionally small (~30 lines of generation code plus a CRD field and webhook validation) and unlocks a meaningful step up in production stability.

**Affected Areas**:
- `operators/keystone/api/v1alpha1/keystone_types.go`
- `operators/keystone/api/v1alpha1/keystone_webhook.go`
- `operators/keystone/internal/controller/reconcile_deployment.go`
- `operators/keystone/internal/controller/reconcile_deployment_test.go`
- `operators/keystone/helm/keystone-operator/crds/keystone.openstack.c5c3.io_keystones.yaml`
- `operators/keystone/config/crd/bases/keystone.openstack.c5c3.io_keystones.yaml`
- `architecture/docs/04-architecture/04-high-availability.md`

**Acceptance Criteria**:
- [ ] `KeystoneSpec` gains an optional `priorityClassName string` field
- [ ] `KeystoneSpec` gains an optional `topologySpreadConstraints []corev1.TopologySpreadConstraint` field
- [ ] `buildKeystoneDeployment` sets default `TopologySpreadConstraints` of `maxSkew: 1` over `topology.kubernetes.io/zone` and `kubernetes.io/hostname`, both `whenUnsatisfiable: ScheduleAnyway`, with `LabelSelector` matching the Deployment's selector labels
- [ ] When `spec.topologySpreadConstraints` is set, it fully replaces the defaults (no merging surprises)
- [ ] When `spec.priorityClassName` is set, it is applied to the pod template; otherwise the field is left empty so admission control does not reject pods outside `kube-system`
- [ ] The validating webhook rejects a `priorityClassName` that does not exist in the cluster (lookup against `scheduling.k8s.io/v1 PriorityClass`) and rejects topology constraints whose label selector does not match the Deployment selector
- [ ] Unit tests in `reconcile_deployment_test.go` cover: defaults applied, override replaces defaults, `priorityClassName` set/unset, webhook validation paths
- [ ] CRD manifests in both `helm/.../crds/` and `config/crd/bases/` are regenerated and committed
- [ ] `architecture/docs/04-architecture/04-high-availability.md` documents the new defaults and the override fields, including a recommended `PriorityClass` definition operators should ship alongside Keystone for production

**Labels:** enhancement

---
Source: https://github.com/C5C3/forge/issues/207

### CODE REVIEW

**Comment by berendt** [comment]
> Reviewed by planwerk-review 3a5c8bd-dirty with Claude CLI



## WARNING (7)

### W-001: Missing E2E Tests: topologySpreadConstraints and priorityClassName
**File**: `tests/e2e/keystone/` — **Fix**: ASK — **Confidence**: verified — **Pattern**: Enforce consistent error-handling pattern across all test cases

**Problem**: The project has 26 Chainsaw E2E test directories under tests/e2e/keystone/, one for every existing feature. This PR adds two new spec fields with non-trivial behavior (default injection, override, cluster-scoped PriorityClass lookup) but includes no E2E test directory, breaking the established 1 feature = 1 E2E directory convention.

**Code**:
```
tests/e2e/keystone/autoscaling/
tests/e2e/keystone/basic-deployment/
tests/e2e/keystone/network-policy/
tests/e2e/keystone/resources/
tests/e2e/keystone/uwsgi/
... (26 total, 0 for topology-spread or priority-class)
```

**Action Required**: Create tests/e2e/keystone/topology-spread/ with a chainsaw-test.yaml that: 1) Applies a Keystone CR with default TSC (nil) and asserts the Deployment has 2 constraints, 2) Patches with custom TSC and asserts verbatim pass-through, 3) Patches with empty slice and asserts no constraints.

**Suggested Fix**:
```
Create tests/e2e/keystone/topology-spread/ and tests/e2e/keystone/priority-class/ directories with chainsaw-test.yaml files covering default injection, custom override, and empty/unset scenarios for each feature.
```

**Related**: W-002

---

### W-002: Missing Tests: TestValidateCreate_RunsAllValidations does not cover new validation paths
**File**: `keystone_webhook_test.go:1107-1167` — **Fix**: AUTO-FIX — **Confidence**: verified — **Pattern**: Ensure test names accurately describe their coverage scope

**Problem**: TestValidateCreate_RunsAllValidations exercises every validation rule simultaneously to prove error accumulation (no short-circuiting). It currently covers 12 validation areas but omits the two new ones: PriorityClassName and TopologySpreadConstraints.

**Code**:
```
func TestValidateCreate_RunsAllValidations(t *testing.T) {
    // ... 12 areas broken ...
    _, err := w.ValidateCreate(context.Background(), k)
    // ... 12 substring assertions ...
    // Missing: priorityClassName and topologySpreadConstraints
}
```

**Action Required**: Add PriorityClassName (nonexistent class) and TopologySpreadConstraints (wrong labels) violations to the test, plus corresponding substring assertions. The PriorityClassName validation requires a Client, so the test needs updating to inject a newFakeClient().Build() into w.

**Suggested Fix**:
```
Add to the test setup: a fake Client on the webhook, a nonexistent PriorityClassName on the Keystone spec, and TopologySpreadConstraints with incorrect label selectors. Add assertions: g.Expect(errMsg).To(ContainSubstring("priorityClassName")) and g.Expect(errMsg).To(ContainSubstring("topologySpreadConstraints")).
```

**Related**: W-001

---

### W-003: Hardcoded labels in TSC validation will drift from selectorLabels()
**File**: `keystone_webhook.go:397-400` — **Fix**: ASK — **Confidence**: likely — **Pattern**: Avoid hardcoded lists that duplicate a dynamic source

**Problem**: The webhook validation hardcodes the expected label map, duplicating the logic in reconcile_deployment.go:146-152 (selectorLabels()). If selectorLabels() ever changes (e.g., adding app.kubernetes.io/component), the webhook will reject valid labels or accept invalid ones.

**Code**:
```
expectedLabels := map[string]string{
    "app.kubernetes.io/name":     "keystone",
    "app.kubernetes.io/instance": k.Name,
}
```

**Action Required**: Either extract the label keys into exported constants in the v1alpha1 package that both the webhook and controller reference, or add a comment cross-referencing reconcile_deployment.go:selectorLabels() to flag the sync requirement.

**Suggested Fix**:
```
Option A: Extract label keys as exported constants in api/v1alpha1/constants.go and reference them from both webhook and controller. Option B: Add a cross-reference comment above the hardcoded map: // Must match selectorLabels() in internal/controller/reconcile_deployment.go
```

**Related**: W-004

---

### W-004: Webhook TSC validation only checks MatchLabels, not MatchExpressions
**File**: `keystone_webhook.go:410` — **Fix**: ASK — **Confidence**: likely

**Problem**: The TSC label selector validation uses maps.Equal(tsc.LabelSelector.MatchLabels, expectedLabels) but does not reject a LabelSelector that adds MatchExpressions. A user could submit a TSC with correct MatchLabels but also include MatchExpressions that widen or narrow the selector beyond the Deployment's intent — the webhook would accept it.

**Code**:
```
if !maps.Equal(tsc.LabelSelector.MatchLabels, expectedLabels) {
    // Rejects wrong MatchLabels, but accepts any MatchExpressions
```

**Action Required**: Add a check that len(tsc.LabelSelector.MatchExpressions) == 0 when the intention is exact-match only, or document that MatchExpressions are intentionally allowed for advanced use cases.

**Suggested Fix**:
```
Add after the MatchLabels check: if len(tsc.LabelSelector.MatchExpressions) > 0 { allErrs = append(allErrs, field.Invalid(..., "topologySpreadConstraints[].labelSelector.matchExpressions must be empty")) }
```

**Related**: W-003

---

### W-005: PriorityClassName validation silently degrades when webhook has no Client
**File**: `keystone_webhook.go:377` — **Fix**: ASK — **Confidence**: uncertain — **Pattern**: Distinguish connectivity failures from valid negative responses

**Problem**: When Client is nil (e.g., unit tests using &KeystoneWebhook{}), the PriorityClassName validation is silently skipped — a nonexistent PriorityClass name is accepted. The test TestValidate_PriorityClassNameEmptyAccepted at line 1503 uses a nil-Client webhook, meaning it tests the nil-Client bypass rather than purely the empty-name logic, which is slightly misleading.

**Code**:
```
if k.Spec.PriorityClassName != nil && *k.Spec.PriorityClassName != "" && w.Client != nil {
```

**Action Required**: No code change needed for the production path. Rename or add a comment to the test at line 1503 to clarify it tests the nil-Client path, not just the empty-name path.

**Suggested Fix**:
```
Rename the test to TestValidate_PriorityClassNameNilOrEmpty_SkipsValidation or add a comment clarifying the nil-Client bypass behavior.
```

---

### W-006: TopologySpreadConstraints and PriorityClassName not propagated to Jobs
**File**: `` — **Fix**: ASK — **Confidence**: uncertain

**Problem**: The new fields are applied to the Deployment PodSpec but not to Job-based workloads (db-sync, bootstrap, fernet rotation, credential rotation). While Jobs are short-lived and topology spread is less critical, PriorityClassName on Jobs matters on resource-constrained clusters where a db-sync Job could be evicted by lower-priority workloads.

**Action Required**: Consider propagating PriorityClassName to Job PodSpecs. TopologySpreadConstraints can be omitted for single-replica Jobs. If intentionally excluded, add a comment documenting why.

**Suggested Fix**:
```
Propagate PriorityClassName to Job PodSpecs in the reconcile functions for db-sync, bootstrap, fernet rotation, and credential rotation. If intentionally excluded, document the rationale in a code comment.
```

---

### W-007: Missing Requirement: architecture HA documentation not verifiable
**File**: `` — **Fix**: ASK — **Confidence**: uncertain

**Problem**: The acceptance criteria requires architecture/docs/04-architecture/04-high-availability.md to document the new defaults and override fields, including a recommended PriorityClass definition. The architecture submodule pointer was updated, but the actual documentation content cannot be verified from this diff since it is a submodule reference.

**Action Required**: Verify the submodule content at the referenced commit includes the required HA documentation updates covering default TopologySpreadConstraints, override behavior, and recommended PriorityClass definition.

**Suggested Fix**:
```
Manually verify the architecture submodule at the referenced commit contains the required documentation. If missing, add the documentation before merge.
```

---

## INFO (1)

### I-001: Shared LabelSelector pointer between default TSC entries
**File**: `reconcile_deployment.go:370-384` — **Fix**: ASK — **Confidence**: verified

**Problem**: Both default TopologySpreadConstraints share the same LabelSelector pointer. This is safe because the PodSpec is consumed by Kubernetes and never mutated in-process, and DeepCopy handles each constraint independently.

**Code**:
```
ls := &metav1.LabelSelector{MatchLabels: selectorLabels(keystone)}
return []corev1.TopologySpreadConstraint{
    {
        // ...
        LabelSelector: ls,
    },
    {
        // ...
        LabelSelector: ls,
    },
}
```

**Action Required**: No action needed. Safe as-is.

**Suggested Fix**:
```
No change required. The shared pointer is safe in this context.
```

---

## Summary

The PR is well-structured with clean implementation, proper CRD regeneration in both locations, RBAC rules added to shared templates, and thorough unit test coverage for webhook validation and deployment builder logic. The most significant gaps are missing E2E/Chainsaw tests (breaking the project's 1-feature-1-directory convention), an incomplete aggregate validation test, and TSC label validation that ignores MatchExpressions. Production code is solid but test coverage needs strengthening before merge.

| Severity | Count |
|----------|-------|
| BLOCKING | 0 |
| CRITICAL | 0 |
| WARNING  | 7 |
| INFO     | 1 |

> [!IMPORTANT]
> **Recommendation**: Merge after addressing the test coverage gaps: add E2E Chainsaw tests for topology-spread and priority-class features, update TestValidateCreate_RunsAllValidations to include the two new validation paths, and decide on MatchExpressions handling in TSC validation. The remaining findings (label duplication, Job propagation, submodule docs) can be addressed in follow-up work if the team agrees.

File: operators/keystone/api/v1alpha1/keystone_types.go [hunk 1]
```diff
@@ -117,6 +117,21 @@ type KeystoneSpec struct {
...
+	// +optional
+	TopologySpreadConstraints []corev1.TopologySpreadConstraint `json:"topologySpreadConstraints,omitempty"`
+
+	// PriorityClassName sets the priority class for Keystone API pods (CC-0075).
+	// When set, the operator passes the value through to the PodSpec, allowing
+	// cluster administrators to control scheduling priority and preemption.
+	// When unset, no priority class is configured and the cluster default applies.
+	// +optional
+	PriorityClassName *string `json:"priorityClassName,omitempty"`
+
```
Comment (gndrmnn) [inline operators/keystone/api/v1alpha1/keystone_types.go]: Changes to the types files require a regeneration of the autogenerated files

**Review by gndrmnn — APPROVED** [review]
